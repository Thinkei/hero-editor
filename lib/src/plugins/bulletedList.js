import React from 'react';
import { Transforms } from 'slate';
import { useSlate } from 'slate-react';
import {
  createPlugin,
  isBlockActive,
  makeRenderElement,
  addMessageListener,
  postMessage,
  isList,
} from '../helpers';
import Toolbar from '../components/Toolbar';
import Icon from '../components/Icon';
import listItem from './listItem';
import { BULLETED_LIST } from '../constants';

const renderElement = makeRenderElement(BULLETED_LIST, 'ul');

const handleMessage = addMessageListener(BULLETED_LIST, ({ editor }) => {
  const isActive = isBlockActive(editor, BULLETED_LIST);

  // Unwrap all the items of any list
  Transforms.unwrapNodes(editor, {
    match: isList,
    split: true,
  });

  if (isActive) {
    Transforms.setNodes(editor, { type: 'paragraph' });
  } else {
    Transforms.setNodes(editor, { type: listItem.name });
    Transforms.wrapNodes(editor, { type: BULLETED_LIST, children: [] });
  }
});

const ToolbarButton = () => {
  const editor = useSlate();

  return (
    <Toolbar.Button
      active={isBlockActive(editor, BULLETED_LIST)}
      onMouseDown={(event) => {
        event.preventDefault();
        postMessage(BULLETED_LIST, {}, editor);
      }}
    >
      <Icon>format_list_bulleted</Icon>
    </Toolbar.Button>
  );
};

export default createPlugin({
  name: BULLETED_LIST,
  renderElement,
  handleMessage,
  ToolbarButton,
});
