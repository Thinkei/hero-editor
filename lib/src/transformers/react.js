import React from 'react';
import { Text, Node } from 'slate';
import y from '../helpers/why';

const isRoot = Array.isArray;

const makeReactTransformer = (rules) =>
  y((reactTransformer) => (node, key) => {
    if (isRoot(node)) return node.map(reactTransformer);

    if (Text.isText(node)) return rules.leaf(node, key);

    const children = node.children.map(reactTransformer);

    return rules.element(node, children, key);
  });

const defaultRules = {
  leaf: (node, key) => {
    let children = node.text;

    if (node.bold) children = <strong key={key}>{children}</strong>;

    if (node.italic) children = <em key={key}>{children}</em>;

    if (node.underline) children = <u key={key}>{children}</u>;

    return <span key={key}>{children}</span>;
  },

  element: (node, children, key) => {
    switch (node.type) {
      case 'paragraph':
        return (
          <p
            key={key}
            style={{ margin: 0, lineHeight: 1.5, minHeight: '1.5em' }}
          >
            {children}
          </p>
        );

      case 'heading-one':
        return <h1 key={key}>{children}</h1>;

      case 'link':
        return (
          <a key={key} href={Node.string(node)} target="_blank">
            {children}
          </a>
        );

      case 'mention':
        return (
          <span
            key={key}
            style={{
              padding: '2px 4px',
              background: '#eeeeee',
              borderRadius: 4,
            }}
          >
            @{node.data.name}
          </span>
        );

      case 'bulleted-list':
        return <ul key={key}>{children}</ul>;

      case 'numbered-list':
        return <ol key={key}>{children}</ol>;

      case 'list-item':
        return <li key={key}>{children}</li>;

      default:
        return <div key={key}>{children}</div>;
    }
  },
};

const defaultReactTransformer = makeReactTransformer(defaultRules);

export default makeReactTransformer;

export { makeReactTransformer, defaultReactTransformer };
