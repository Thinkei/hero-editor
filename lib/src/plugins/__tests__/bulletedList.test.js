import { createEditor } from 'slate';
import { render, waitFor } from '@testing-library/react';
import bulletedList from '../bulletedList';
import {
  createPlugin,
  renderDefaultElement,
  postMessage,
  withId,
} from '../../helpers';

describe('bulletedList plugin', () => {
  it('is a plugin object', () => {
    for (const prop in createPlugin()) {
      expect(bulletedList()[prop]).toBeDefined();
    }
  });

  it('renders ul element', () => {
    const { baseElement } = render(
      bulletedList().renderElement(renderDefaultElement)({
        element: { type: 'bulleted-list' },
        children: 'Bulleted list',
      }),
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('handles bulleted-list message properly', async () => {
    const editor = withId('sample')(createEditor());
    bulletedList().handleMessage(editor);
    editor.insertNode({
      type: 'paragraph',
      children: [
        {
          text: 'A line of text',
        },
      ],
    });
    postMessage('bulleted-list', {}, editor);
    await waitFor(() => {
      expect(editor.children).toEqual([
        {
          type: 'bulleted-list',
          children: [
            {
              type: 'list-item',
              children: [
                {
                  text: 'A line of text',
                },
              ],
            },
          ],
        },
      ]);
    });
  });
});
