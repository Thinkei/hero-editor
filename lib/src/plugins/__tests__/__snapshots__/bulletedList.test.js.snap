// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin bulleted list renders ul element 1`] = `
<body>
  <div>
    <ul>
      Bulleted list
    </ul>
  </div>
</body>
`;
